src/client/index.ts

```diff
@@ -527,64 +597,158 @@ export class Client<
         return this.request({ method: 'resources/unsubscribe', params }, EmptyResultSchema, options);
     }
 
-    async callTool(
+    /**
+     * Calls a tool and waits for the result. Automatically validates structured output if the tool has an outputSchema.
+     *
+     * For task-based execution with streaming behavior, use callToolStream() instead.
+     */
+    async callTool<T extends typeof CallToolResultSchema | typeof CompatibilityCallToolResultSchema>(
         params: CallToolRequest['params'],
-        resultSchema: typeof CallToolResultSchema | typeof CompatibilityCallToolResultSchema = CallToolResultSchema,
+        resultSchema: T = CallToolResultSchema as T,
         options?: RequestOptions
-    ) {
-        const result = await this.request({ method: 'tools/call', params }, resultSchema, options);
+    ): Promise<SchemaOutput<T>> {
+        return await takeResult(this.callToolStream<T>(params, resultSchema, options));
+    }
 
-        // Check if the tool has an outputSchema
+    /**
+     * Calls a tool and returns an AsyncGenerator that yields response messages.
+     * The generator is guaranteed to end with either a 'result' or 'error' message.
+     *
+     * This method provides streaming access to tool execution, allowing you to
+     * observe intermediate task status updates for long-running tool calls.
+     * Automatically validates structured output if the tool has an outputSchema.
+     *
+     * For simple tool calls without streaming, use callTool() instead.
+     *
+     * @example
+     * ```typescript
+     * const stream = client.callToolStream({ name: 'myTool', arguments: {} });
+     * for await (const message of stream) {
+     *   switch (message.type) {
+     *     case 'taskCreated':
+     *       console.log('Tool execution started:', message.task.taskId);
+     *       break;
+     *     case 'taskStatus':
+     *       console.log('Tool status:', message.task.status);
+     *       break;
+     *     case 'result':
+     *       console.log('Tool result:', message.result);
+     *       // Structured output is automatically validated
+     *       break;
+     *     case 'error':
+     *       console.error('Tool error:', message.error);
+     *       break;
+     *   }
+     * }
+     * ```
+     *
+     * @param params - Tool call parameters (name and arguments)
+     * @param resultSchema - Zod schema for validating the result (defaults to CallToolResultSchema)
+     * @param options - Optional request options (timeout, signal, task creation params, etc.)
+     * @returns AsyncGenerator that yields ResponseMessage objects
+     */
+    async *callToolStream<T extends typeof CallToolResultSchema | typeof CompatibilityCallToolResultSchema>(
```

Comment:
do we expect end users to actually call this function?

If so probably worth having it in the README.